compiler/optimizer/OSRDefAnalysis.cpp

```diff
@@ -699,75 +699,280 @@ TR_OSRLiveRangeAnalysis::TR_OSRLiveRangeAnalysis(TR::OptimizationManager *manage
+         TR_ASSERT(osrPoint != NULL, "Cannot find a pre OSR point for node %p", node);
+
+         TR_BitVector *ppInfo = osrMethodData->getPendingPushLivenessInfo(bci.getByteCodeIndex());
+         pendingPushLiveRangeInfo(node, ppInfo, _pendingPushSymRefs, osrPoint, osrMethodData);
+         pendingPushSlotSharingInfo(node, ppInfo, _sharedSymRefs, osrPoint);
          }
-      return 0;
       }
-   else
+
+   // Certain modes require the method entry to be an OSR point
+   if (comp()->isOutermostMethod() && comp()->getHCRMode() == TR::osr)
       {
-      if (comp()->getOption(TR_TraceOSR))
+      TR_ByteCodeInfo bci;
+      bci.setCallerIndex(-1);
+      bci.setByteCodeIndex(0);
+      TR_OSRPoint *osrPoint = comp()->getMethodSymbol()->findOSRPoint(bci);
+      TR_ASSERT(osrPoint != NULL, "Cannot find a OSR point for method entry");
+
+      // There should be no slot sharing at the method entry
+      pendingPushLiveRangeInfo(comp()->getStartTree()->getNode(), NULL, _pendingPushSymRefs, osrPoint, osrMethodData);
+      pendingPushSlotSharingInfo(comp()->getStartTree()->getNode(), NULL, _sharedSymRefs, osrPoint);
+      }
+
+   return 1;
+   }
+
+/*
+ * Allocate the BitVector of dead symbol references against the provided OSR point.
+ */
+void TR_OSRLiveRangeAnalysis::pendingPushLiveRangeInfo(TR::Node *node, TR_BitVector *liveSymRefs,
+   TR_BitVector &allPendingPushSymRefs, TR_OSRPoint *osrPoint, TR_OSRMethodData *osrMethodData)
+   {
+   int32_t byteCodeIndex = osrPoint->getByteCodeInfo().getByteCodeIndex();
+   TR_BitVector *deadBV = NULL;
+
+   // Calculate the dead symbol references, allocating a new BitVector if its required
+   _workBitVector = allPendingPushSymRefs;
+   if (liveSymRefs)
+      _workBitVector -= *liveSymRefs;
+   if (!_workBitVector.isEmpty())
+      {
+      deadBV = new (trHeapMemory()) TR_BitVector(0, trMemory(), heapAlloc);
+      *deadBV = _workBitVector;
+      osrMethodData->addLiveRangeInfo(byteCodeIndex, deadBV);
+      }
+
+   if (comp()->getOption(TR_TraceOSR))
+      {
+      traceMsg(comp(), "Live PP variables at OSR point %p of %p bytecode offset %d\n",
+         node, osrMethodData, byteCodeIndex);
+      if (!liveSymRefs)
+         traceMsg(comp(), " NULL");
+      else
+         liveSymRefs->print(comp());
+      traceMsg(comp(), "\n");
+      }
+   }
+
+/*
+ * Solve slot sharing based on the pending push symrefs
+ */
+void TR_OSRLiveRangeAnalysis::pendingPushSlotSharingInfo(TR::Node *node, TR_BitVector *liveSymRefs,
+   TR_BitVector &slotSharingSymRefs, TR_OSRPoint *osrPoint)
+   {
+   if (liveSymRefs && !liveSymRefs->isEmpty())
+      {
+      TR::SymbolReferenceTable *symRefTab = comp()->getSymRefTab();
+      TR_ByteCodeInfo& bcInfo = osrPoint->getByteCodeInfo();
+
+      if (trace())
+         traceMsg(comp(), "Shared PP slots at OSR point [%p] at %d:%d\n", node, bcInfo.getCallerIndex(), bcInfo.getByteCodeIndex());
+
+      _workBitVector = *liveSymRefs;
```

Comment:
Not sure about the allocator used with this copy - I feel something else would make the lifetime of _workBitVector more explicit...